controllers/kafkauser_controller.go

```diff
@@ -332,8 +332,10 @@ func (r *KafkaUserReconciler) checkFinalizers(ctx context.Context, cluster *v1be
 	var err error
 	if util.StringSliceContains(instance.GetFinalizers(), userFinalizer) {
 		if len(instance.Spec.TopicGrants) > 0 {
-			if err = r.finalizeKafkaUserACLs(reqLogger, cluster, user); err != nil {
-				return requeueWithError(reqLogger, "failed to finalize kafkauser", err)
+			for _, topicGrant := range instance.Spec.TopicGrants {
+				if err = r.finalizeKafkaUserACLs(reqLogger, cluster, user, topicGrant.PatternType); err != nil {
```

Comment:
Passing `topicGrant.PatternType` directly could be problematic due to these few lines in the `CreateUserACLs` function: https://github.com/banzaicloud/koperator/blob/master/pkg/kafkaclient/users.go#L45-L47

If the `topicGrant.PatternType` is empty, then when it gets to:
```golang
aclPatternType := AclPatternTypeMapping(patternType)
```
the `aclPatternType` will be `sarama.AclPatternUnknown`

And at the end when the delete ACL request gets to the Kafka cluster, the request will fail in the validation phase, for example: 

```
[2023-06-08 19:10:59,462] ERROR Exception while processing request from 10.244.4.61:29092-127.0.0.6:44155-237 (kafka.network.Processor)
org.apache.kafka.common.errors.InvalidRequestException: Error getting request for apiKey: DELETE_ACLS, apiVersion: 1, connectionId: 10.244.4.61:29092-127.0.0.6:44155-237, listenerName: ListenerName(INTERNAL), principal: User:CN=kafka-kafka-operator
Caused by: java.lang.IllegalArgumentException: Filters contain UNKNOWN elements, filters: [DeleteAclsFilter(resourceTypeFilter=1, resourceNameFilter=null, patternTypeFilter=0, principalFilter='CN=kafka-user', hostFilter=null, operation=6, permissionType=1)]
```

Therefore, we need the [same if statement](https://github.com/banzaicloud/koperator/blob/master/pkg/kafkaclient/users.go#L45-L47) to make sure the `topicGrant.PatternType` is not empty

Reply:
Nice catch. Thank you for pointing out. Added the suggested line in DeleteUserACLs function instead of adding here, as it is better to take care in the lower function in case it is used somewhere else in future.